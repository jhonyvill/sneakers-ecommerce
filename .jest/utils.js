import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../src/Themes";

export function renderWithTheme(children) {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
}

export const productDefault = {
  id: "id-1",
  brand: "brand mock",
  title: "title mock",
  price: 250,
  discount: 0.5,
  description: "description mock",
  pictures: [
    {
      src: "image-1.jpg",
      thumb: "thumbnail-1.jpg",
      alt: "product 1",
    },
    {
      src: "image-2.jpg",
      thumb: "thumbnail-2.jpg",
      alt: "product 2",
    },
    {
      src: "image-3.jpg",
      thumb: "thumbnail-3.jpg",
      alt: "product 3",
    },
    {
      src: "image-4.jpg",
      thumb: "thumbnail-4.jpg",
      alt: "product 4",
    },
    {
      src: "image-5.jpg",
      thumb: "thumbnail-5.jpg",
      alt: "product 5",
    },
  ],
};

export const itemDefault = {
  product: { ...productDefault },
  quantity: 1,
};
