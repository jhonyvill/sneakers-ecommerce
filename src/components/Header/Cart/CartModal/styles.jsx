import styled from "styled-components";

export const Modal = styled.section`
  width: 21.875rem;
  min-width: 15.625rem;
  min-height: 15rem;

  border: none;
  border-radius: 0.5rem;
  background-color: ${({ theme }) => theme.colors.white};
  box-shadow: 0px 15px 20px -6px ${({ theme }) => theme.colors.grayishBlue};

  font-size: ${({ theme }) => theme.sizes.s};
  font-weight: 700;

  position: absolute;
  top: 6rem;
  right: -1.25rem;
  z-index: 200;

  box-sizing: border-box;

  display: flex;
  flex-direction: column;

  opacity: 0;
  transform: translateY(-1.25rem);
  animation: transformAnimation 0.2s forwards;

  @media (max-width: ${({ theme }) => theme.screen.largeMobile}) {
    right: 0;
    width: calc(100% - 2rem);
    margin: 0 1rem;
  }
`;

export const Title = styled.div`
  padding: 1rem 1.5rem;
  border-bottom: 2px solid ${({ theme }) => theme.colors.lightGrayishBlue};

  & h2 {
    margin: 0;
    padding: 0;
    font-size: ${({ theme }) => theme.sizes.s};
  }
`;

export const Content = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;
  flex-grow: 1;
  padding: 1.5rem;

  & button {
    margin-top: 1.5rem;
  }
`;

export const Items = styled.ul`
  width: 100%;
  max-height: 21.875rem;
  overflow: auto;
  margin: 0;
  display: flex;
  flex-direction: column;
  gap: 0.75rem 0;

  &::-webkit-scrollbar {
    width: 4px;
  }
  &::-webkit-scrollbar-thumb {
    background-color: ${({ theme }) => theme.colors.primaryOrange};
    border-radius: 4px;
  }
`;
