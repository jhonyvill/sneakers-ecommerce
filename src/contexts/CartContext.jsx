import React from "react";
import { useReducer } from "react";

export const CartContext = React.createContext();

export const CartProvider = ({ children }) => {
  const initialState = { items: [] };

  function cartReducer(state, action) {
    switch (action.type) {
      case "ADD_ITEM":
        return { ...state, items: [...state.items, action.payload] };
      case "UPDATE_ITEM":
        const { product } = action.payload;
        const itemIndex = state.items.findIndex(
          (item) => item.product.id === product.id
        );

        const updatedItems = [...state.items];
        updatedItems.slice(itemIndex, product);
        return { ...state, items: updatedItems };
      default:
        state;
    }
  }

  const [state, dispatch] = useReducer(cartReducer, initialState);

  return (
    <>
      <CartContext.Provider
        value={{ cartState: state, cartDispatch: dispatch }}
      >
        {children}
      </CartContext.Provider>
    </>
  );
};
