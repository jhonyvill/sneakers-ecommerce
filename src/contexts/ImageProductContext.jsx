import React from "react";
import { useState } from "react";

export const ImageProductContext = React.createContext();

export const ImageProductProvider = ({ children }) => {
  const [selectedImage, setSelectedImage] = useState({
    image: null,
    index: 0,
  });

  const handleSelectedModalImage = (image, index) => {
    setSelectedImage({ image, index });
  };

  return (
    <>
      <ImageProductContext.Provider
        value={{
          imageSelected: selectedImage.image,
          indexImageSelected: selectedImage.index,
          handleSelectedModalImage,
        }}
      >
        {children}
      </ImageProductContext.Provider>
    </>
  );
};
