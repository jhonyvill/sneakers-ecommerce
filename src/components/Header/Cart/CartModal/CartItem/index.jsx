import React from "react";
import * as S from "./styles";
import TrashSvg from "../../../../../assets/icon-delete.svg?react";
import { useProductPrice } from "../../../../../hooks/useProductPrice";
import { useCart } from "../../../../../hooks/useCart";
import PropTypes from "prop-types";

const CartItem = ({ item }) => {
  const { finalPrice, getTotalItemPrice } = useProductPrice(item.product);
  const totalItemPrice = getTotalItemPrice(item.quantity);
  const { removeCartProduct } = useCart();

  return (
    <S.Item>
      <S.Photo
        src={item.product.pictures[0].thumb}
        alt={item.product.pictures[0].alt}
        width="176px"
        height="176px"
      />
      <S.Title>{item.product.title}</S.Title>
      <S.PriceContainer>
        <span>{`$${finalPrice} x ${item.quantity}`}</span>
        <S.TotalPrice>${totalItemPrice}</S.TotalPrice>
      </S.PriceContainer>
      <S.DeleteButton
        onClick={() => removeCartProduct(item.product.id)}
        aria-label="deletar item"
      >
        <TrashSvg />
      </S.DeleteButton>
    </S.Item>
  );
};

CartItem.propTypes = {
  item: PropTypes.object,
};

export default CartItem;
