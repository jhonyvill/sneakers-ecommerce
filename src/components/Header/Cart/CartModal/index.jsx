import * as S from "./styles";
import { useContext, useEffect } from "react";
import { CartContext } from "../../../../contexts/CartContext";
import PropTypes from "prop-types";
import CartItem from "./CartItem";
import Button from "../../../Utils/Button";

const CartModal = ({ cartRef, isModalOpen, handleCheckout, hasItems }) => {
  const { cartState } = useContext(CartContext);

  function handleClickOutside(event) {
    if (!cartRef.current.contains(event.target)) isModalOpen(false);
  }

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    return () => document.removeEventListener("mousedown", handleClickOutside);
  }, []);

  return (
    <S.Modal aria-label="modal carrinho de compras">
      <S.Title>
        <h2>Cart</h2>
      </S.Title>
      <S.Content>
        {hasItems ? (
          <>
            <S.Items>
              {cartState.items.map((item, index) => (
                <li key={index}>
                  <CartItem item={item} />
                </li>
              ))}
            </S.Items>
            <Button aria-label="finalizar compra" onClick={handleCheckout}>Checkout</Button>
          </>
        ) : (
          <p>Your cart is empty.</p>
        )}
      </S.Content>
    </S.Modal>
  );
};

CartModal.propTypes = {
  cartRef: PropTypes.object,
  closeCartModal: PropTypes.func,
  handleCheckout: PropTypes.func,
  hasItems: PropTypes.bool
};

export default CartModal;
