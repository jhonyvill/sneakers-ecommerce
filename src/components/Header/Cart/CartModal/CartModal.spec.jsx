import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme, itemDefault } from "../../../../../.jest/utils";
import { CartContext } from "../../../../contexts/CartContext";
import CartModal from ".";

function renderCartModal(cartRef, isModalOpen, handleCheckout, hasItems) {
  return renderWithTheme(
    <CartContext.Provider
      value={{ cartState: { items: [itemDefault, itemDefault] } }}
    >
      <CartModal
        cartRef={cartRef}
        isModalOpen={isModalOpen}
        handleCheckout={handleCheckout}
        hasItems={hasItems}
      />
    </CartContext.Provider>
  );
}

describe("<CartModal />", () => {
  it("should render CartModal component empty", () => {
    renderCartModal({}, jest.fn(), jest.fn(), false);

    expect(screen.getByRole("heading", { name: "Cart" })).toBeInTheDocument();
    expect(screen.getByText(/your cart is empty\./i)).toBeInTheDocument();
  });

  it("should render CartModal component with item", () => {
    renderCartModal({}, jest.fn(), jest.fn(), true);

    const checkoutButton = screen.getByRole("button", { name: /finalizar compra/i });

    expect(checkoutButton).toBeInTheDocument();
    expect(checkoutButton).toHaveTextContent("Checkout");
    expect(screen.getByRole("heading", { name: "Cart" })).toBeInTheDocument();
    expect(screen.getByRole("list")).toBeInTheDocument();
    expect(screen.getAllByRole("listitem")).toHaveLength(2);
  });

  it("should close modal when different ref element clicked", async () => {
    const isModalOpenMock = jest.fn();
    const outsideElement = document.createElement("div");

    renderCartModal(
      { current: document.createElement("div") },
      isModalOpenMock,
      jest.fn(),
      false
    );
    document.body.appendChild(outsideElement);

    expect.assertions(3);
    await userEvent.click(outsideElement);
    expect(isModalOpenMock).toHaveBeenCalled();
    expect(isModalOpenMock).toHaveBeenCalledTimes(1);
    expect(isModalOpenMock).toHaveBeenCalledWith(false);
    outsideElement.remove();
  });
  it("shouldn't close modal when ref element clicked", async () => {
    const isModalOpenMock = jest.fn();
    const outsideElement = document.createElement("div");
    const elementRef = document.createElement("div");
    outsideElement.appendChild(elementRef);

    renderCartModal(
      { current: outsideElement },
      isModalOpenMock,
      jest.fn(),
      false
    );
    document.body.appendChild(outsideElement);

    expect.assertions(1);
    await userEvent.click(elementRef);
    expect(isModalOpenMock).not.toHaveBeenCalled();
    outsideElement.remove();
  });

  it("should call checkout function", async () => {
    const handleCheckoutMock = jest.fn();
    renderCartModal(
      { current: document.createElement("div") },
      jest.fn(),
      handleCheckoutMock,
      true
    );

    const checkoutButton = screen.getByRole("button", { name: /finalizar compra/i });

    expect.assertions(1);
    await userEvent.click(checkoutButton);
    expect(handleCheckoutMock).toHaveBeenCalledTimes(1);
  });

});

describe("<CartModal /> snapshots", () => {
  it("should match snapshot", () => {
    const {container} = renderCartModal({}, jest.fn(), jest.fn(), true);
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should match snapshot when cart is empty", () => {
    const {container} = renderCartModal({}, jest.fn(), jest.fn(), false);
    expect(container.firstChild).toMatchSnapshot();
  });
});
