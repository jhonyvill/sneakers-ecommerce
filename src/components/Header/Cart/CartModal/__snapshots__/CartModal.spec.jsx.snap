// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartModal /> snapshots should match snapshot 1`] = `
<section
  aria-label="modal carrinho de compras"
  class="sc-bdnyFh ieuDOa"
>
  <div
    class="sc-gtsqUy bMTtzH"
  >
    <h2>
      Cart
    </h2>
  </div>
  <div
    class="sc-dlniIP kWsVGs"
  >
    <ul
      class="sc-hKFymg jvPhoA"
    >
      <li>
        <div
          class="sc-eCAqax cCMqVm"
        >
          <img
            alt="product 1"
            class="sc-jSFipO dNdahi"
            height="176px"
            src="thumbnail-1.jpg"
            width="176px"
          />
          <h2
            class="sc-gKAaef dLbczZ"
          >
            title mock
          </h2>
          <div
            class="sc-iCoHzw gtVcVg"
          >
            <span>
              $125.00 x 1
            </span>
            <span
              class="sc-fujznN iuzEom"
              data-testid="total-price"
            >
              $
              125.00
            </span>
          </div>
          <button
            aria-label="deletar item"
            class="sc-pNWRh cqTcdp"
          >
            <icon-mock
              data-testid="trash-svg"
            />
          </button>
        </div>
      </li>
      <li>
        <div
          class="sc-eCAqax cCMqVm"
        >
          <img
            alt="product 1"
            class="sc-jSFipO dNdahi"
            height="176px"
            src="thumbnail-1.jpg"
            width="176px"
          />
          <h2
            class="sc-gKAaef dLbczZ"
          >
            title mock
          </h2>
          <div
            class="sc-iCoHzw gtVcVg"
          >
            <span>
              $125.00 x 1
            </span>
            <span
              class="sc-fujznN iuzEom"
              data-testid="total-price"
            >
              $
              125.00
            </span>
          </div>
          <button
            aria-label="deletar item"
            class="sc-pNWRh cqTcdp"
          >
            <icon-mock
              data-testid="trash-svg"
            />
          </button>
        </div>
      </li>
    </ul>
    <button
      aria-label="finalizar compra"
      class="sc-jrsKJM jabYbl"
    >
      Checkout
    </button>
  </div>
</section>
`;

exports[`<CartModal /> snapshots should match snapshot when cart is empty 1`] = `
<section
  aria-label="modal carrinho de compras"
  class="sc-bdnyFh ieuDOa"
>
  <div
    class="sc-gtsqUy bMTtzH"
  >
    <h2>
      Cart
    </h2>
  </div>
  <div
    class="sc-dlniIP kWsVGs"
  >
    <p>
      Your cart is empty.
    </p>
  </div>
</section>
`;
